File: python/test/t_DualLinearCombinationFunction_std.py
#! /usr/bin/env python

from __future__ import print_function
import openturns as ot

# First, build two functions from R^3->R
inVar = ['x1', 'x2', 'x3']
functions = list()
functions.append(ot.SymbolicFunction(
    inVar, ['x1^3 * sin(x2 + 2.5 * x3) - (x1 + x2)^2 / (1.0 + x3^2)']))
functions.append(
    ot.SymbolicFunction(inVar, ['exp(-x1 * x2 + x3) / cos(1.0 + x2 * x3 - x1)']))
# Second, build the weights
coefficients = ot.Sample(0, 3)
coefficients.add([1.5, 2.5, -0.5])
coefficients.add([-3.5, 0.5, -1.5])
# Third, build the function
myFunction = ot.DualLinearCombinationFunction(functions, coefficients)
inPoint = ot.Point([1.2, 2.3, 3.4])
print('myFunction=', myFunction)
print('Value at ', inPoint, '=', myFunction(inPoint))
print('Gradient at ', inPoint, '=', myFunction.gradient(inPoint))
ot.PlatformInfo.SetNumericalPrecision(5)
print('Hessian at ', inPoint, '=', myFunction.hessian(inPoint))
for i in range(myFunction.getOutputDimension()):
    print('Marginal ', i, '=', myFunction.getMarginal(i))
print('Marginal (0,1)=', myFunction.getMarginal([0, 1]))
print('Marginal (0,2)=', myFunction.getMarginal([0, 2]))
print('Marginal (1,2)=', myFunction.getMarginal([1, 2]))

File: lib/test/t_DualLinearCombinationFunction_std.cxx
//                                               -*- C++ -*-
/**
 *  @brief The test file of class Function for dual_linear combinations
 *
 *  Copyright 2005-2022 Airbus-EDF-IMACS-ONERA-Phimeca
 *
 *  This library is free software: you can redistribute it and/or modify
 *  it under the terms of the GNU Lesser General Public License as published by
 *  the Free Software Foundation, either version 3 of the License, or
 *  (at your option) any later version.
 *
 *  This library is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU Lesser General Public License for more details.
 *
 *  You should have received a copy of the GNU Lesser General Public License
 *  along with this library.  If not, see <http://www.gnu.org/licenses/>.
 *
 */
#include "openturns/OT.hxx"
#include "openturns/OTtestcode.hxx"

using namespace OT;
using namespace OT::Test;

int main(int, char *[])
{
  TESTPREAMBLE;
  OStream fullprint(std::cout);

  try
  {
    // First, build two functions from R^3->R
    Description inVar(3);
    inVar[0] = "x1";
    inVar[1] = "x2";
    inVar[2] = "x3";
    Description formula(1);
    formula[0] = "x1^3 * sin(x2 + 2.5 * x3) - (x1 + x2)^2 / (1.0 + x3^2)";
    DualLinearCombinationEvaluation::FunctionCollection functions(2);
    functions[0] = SymbolicFunction(inVar, formula);
    formula[0] = "exp(-x1 * x2 + x3) / cos(1.0 + x2 * x3 - x1)";
    functions[1] = SymbolicFunction(inVar, formula);
    // Second, build the weights
    Sample coefficients(0, 3);
    Point p(3);
    p[0] = 1.5;
    p[1] = 2.5;
    p[2] = -0.5;
    coefficients.add(p);
    p[0] = -3.5;
    p[1] = 0.5;
    p[2] = -1.5;
    coefficients.add(p);
    // Third, build the function
    DualLinearCombinationFunction myFunction(functions, coefficients);
    Point inPoint(3);
    inPoint[0] = 1.2;
    inPoint[1] = 2.3;
    inPoint[2] = 3.4;
    fullprint << "myFunction=" << myFunction << std::endl;
    fullprint << "Value at " << inPoint << "=" << myFunction(inPoint) << std::endl;
    fullprint << "Gradient at " << inPoint << "=" << myFunction.gradient(inPoint) << std::endl;
    PlatformInfo::SetNumericalPrecision(5);
    fullprint << "Hessian at " << inPoint << "=" << myFunction.hessian(inPoint) << std::endl;
    for (UnsignedInteger i = 0; i < myFunction.getOutputDimension(); ++i)
    {
      fullprint << "Marginal " << i << "=" << myFunction.getMarginal(i) << std::endl;
    }
    Indices indices(2);
    indices[0] = 0;
    indices[1] = 1;
    fullprint << "Marginal (0,1)=" << myFunction.getMarginal(indices) << std::endl;
    indices[0] = 0;
    indices[1] = 2;
    fullprint << "Marginal (0,2)=" << myFunction.getMarginal(indices) << std::endl;
    indices[0] = 1;
    indices[1] = 2;
    fullprint << "Marginal (1,2)=" << myFunction.getMarginal(indices) << std::endl;
  }
  catch (TestFailed & ex)
  {
    std::cerr << ex << std::endl;
    return ExitCode::Error;
  }


  return ExitCode::Success;
}
